core/src/main/java/com/datastax/oss/driver/api/core/paging/Pager.java

```
@@ -0,0 +1,324 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.driver.api.core.paging;
+
+import com.datastax.oss.driver.api.core.AsyncPagingIterable;
+import com.datastax.oss.driver.api.core.PagingIterable;
+import com.datastax.oss.driver.shaded.guava.common.collect.ImmutableList;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+
+public class Pager {
+
+  /** A page returned as the result of an offset query. */
+  public interface Page<ElementT> {
+
+    /** The elements in the page. */
+    @NonNull
+    List<ElementT> getElements();
+
+    /**
+     * The page number.
+     *
+     * <p>It will usually match the requested page number, except if the requested page was past the
+     * end of the iterable, and the out of bounds strategy is {@link
+     * OutOfBoundsStrategy#RETURN_LAST_PAGE}.
+     */
+    int getPageNumber();
+
+    /** Whether there are other pages after this one. */
+    boolean isLast();
+  }
+
+  /** What to do if the caller requests a page that is past the end of the current result set. */
+  public enum OutOfBoundsStrategy {
+    /**
+     * Fail with an {@link IndexOutOfBoundsException}.
+     *
+     * <p>As a special case, if the result set is empty and the first page is requested, this
+     * strategy will return an empty page, not fail.
+     */
+    FAIL,
```

I wonder if it's even worth exposing this strategy at all. Its utility is dubious, and it can always be emulated by using `RETURN_EMPTY_PAGE` and throwing if `page.getPageNumber() > 1 && page.getElements().isEmpty()`.

---

I lean into the direction of not providing `OutOfBoundsStrategy` and let the caller handle those cases. We could throw an exception in that case and the user can catch it and returns empty(or last page) if he needs to. Was this `OutOfBoundsStrategy` a feature request? I have the impression that supporting this complicates the implementation a lot. wdyt?

---

If we only keep one strategy, it should be `RETURN_LAST_PAGE`, it's the only one that can't be implemented in term of the others.

Now that I think of it, `RETURN_EMPTY_PAGE` is kind of whacky, I don't think it would be very common in a UI or REST API.